cluster/plan.go

```diff
@@ -770,11 +771,41 @@ func (c *Cluster) BuildEtcdProcess(host *hosts.Host, etcdHosts []*hosts.Host, pr
 	registryAuthConfig, _, _ := docker.GetImageRegistryConfig(c.Services.Etcd.Image, c.PrivateRegistriesMap)
 
 	Env := []string{}
-	Env = append(Env, "ETCDCTL_API=3")
-	Env = append(Env, fmt.Sprintf("ETCDCTL_ENDPOINT=https://%s:2379", listenAddress))
-	Env = append(Env, fmt.Sprintf("ETCDCTL_CACERT=%s", pki.GetCertPath(pki.CACertName)))
-	Env = append(Env, fmt.Sprintf("ETCDCTL_CERT=%s", pki.GetCertPath(nodeName)))
-	Env = append(Env, fmt.Sprintf("ETCDCTL_KEY=%s", pki.GetKeyPath(nodeName)))
+	etcdTag, err := util.GetImageTagFromImage(c.Services.Etcd.Image)
+	if err != nil {
+		logrus.Warn(err)
+	}
+	etcdSemVer, err := util.StrToSemVer(etcdTag)
+	if err != nil {
+		logrus.Warn(err)
+	}
+	etcdChangedEnvSemVer, err := util.StrToSemVer(EtcdChangedEnvVersion)
+	if err != nil {
+		logrus.Warn(err)
+	}
+
+	// Keep the same configuration to avoid updating etcd
+	if etcdSemVer.LessThan(*etcdChangedEnvSemVer) {
+		logrus.Debugf("Version [%s] is less than version [%s]", etcdSemVer, etcdChangedEnvSemVer)
+		Env = append(Env, "ETCDCTL_API=3")
+		Env = append(Env, fmt.Sprintf("ETCDCTL_ENDPOINT=https://%s:2379", listenAddress))
+		Env = append(Env, fmt.Sprintf("ETCDCTL_CACERT=%s", pki.GetCertPath(pki.CACertName)))
+		Env = append(Env, fmt.Sprintf("ETCDCTL_CERT=%s", pki.GetCertPath(nodeName)))
+		Env = append(Env, fmt.Sprintf("ETCDCTL_KEY=%s", pki.GetKeyPath(nodeName)))
+	} else {
+		logrus.Debugf("Version [%s] is equal or higher than version [%s]", etcdSemVer, etcdChangedEnvSemVer)
+		Env = append(Env, "ETCDCTL_API=3")
+		// Point etcdctl to localhost in case we have listen all (0.0.0.0) configured
+		if listenAddress == "0.0.0.0" {
+			Env = append(Env, "ETCDCTL_ENDPOINTS=https://127.0.0.1:2379")
+			// If internal address is configured, set endpoint to that address as well
+		} else {
+			Env = append(Env, fmt.Sprintf("ETCDCTL_ENDPOINTS=https://%s:2379", listenAddress))
+		}
+		Env = append(Env, fmt.Sprintf("ETCDCTL_CACERT=%s", pki.GetCertPath(pki.CACertName)))
```

Comment:
Can we not repeat the common environments including:

ETCDCTL_API
ETCDCTL_CACERT
ETCDCTL_CERT
ETCDCTL_KEY

Reply:
I thought about this and that would mean that to keep the correct order (to not change anything to trigger an update):

ETCDCTL_API first, then the if block, and then the rest, which makes it less readable in my opinion than this. Let me know